code/game/objects/obj_construction.dm

```diff
@@ -0,0 +1,292 @@
+				deconstruction_message = "unscrewing the legs from \the"\
+			)
+		
+		/obj/chair/OnDeconstruction(state_id, mob/user, forced)
+			if(forced)
+				user << "You break \the [src] apart"
+			if(!state_id)
+				qdel(src)
+		
+		/obj/item/wood/attack_self()	// build the base of the chair
+			var/obj/C = new /obj/chair(get_turf(src))
+			C.Construct()
+		
+		
+*/
+
+/datum/construction_state
+	var/id  //states are sequenced from deconstructed -> 1 -> 2 -> 3 -> fully constructed
+
+	/*
+		These can be:
+			null for requiring attack_self/attack_hand on help intent
+			An /obj/item path for a required tool
+			An /obj/item/stack path for required materials (construction only)
+
+		If it's the last one, required_amount_to_construct is the amount of that material required to reach the next state
+		Used materials will be extracted when the object is deconstructed to this state or if the user attack_hand's on help intent
+	*/
+	var/required_type_to_construct
+	var/required_type_to_deconstruct
+
+	var/required_amount_to_construct
+
+	var/construction_delay      //number multiplied by toolspeed, hands and materials have a pseudo-toolspeed of 1
+	var/deconstruction_delay
+
+	var/construction_message    //message displayed in the format user.visible_message("[user] [message]", "You [message]")
+	var/deconstruction_message
+
+	var/examine_message	//null values do not adjust these
+	var/icon_state
+
+	var/max_integrity   //null values do not adjust these, these should always be smaller for earlier construction steps
+	var/failure_integrity
+
+	var/damage_reachable    //if the object can be deconstructed into this state from the next through breaking the object
+							//note that if this is TRUE, modify_max_integrity will not be used to change the object's max_integrity
+							//when constructed from or deconstructed into this state. The change will be purely additive
+							//Having this be TRUE also means that OnConstruction and OnDeconstruction may have null user parameters
+
+	var/datum/construction_state/next_state		//the state that will be next once constructed
+	var/datum/construction_state/prev_state		//the state that will be next once deconstructed
+
+/datum/construction_state/first/New(obj/parent, material_type, material_amount)
+	..(parent, material_type, material_amount)
+
+/datum/construction_state/last/New(obj/parent, required_type_to_deconstruct, deconstruction_delay = 0, deconstruction_message)
+	..(parent, -1 /*don't want to construct with attack_hand/self*/, null, required_type_to_deconstruct, 0, deconstruction_delay, null, deconstruction_message)
+
+/datum/construction_state/New(obj/parent, required_type_to_construct, required_amount_to_construct, required_type_to_deconstruct, construction_delay = 0, deconstruction_delay = 0,\
+ 								construction_message, deconstruction_message, examine_message, icon_state, max_integrity, failure_integrity, damage_reachable = FALSE)
+	src.required_type_to_construct = required_type_to_construct

```

Comment:
Reee. This is a legitimate case of needing `src`, but shouldn't proc argument names be avoided that *require* the use of `src`?